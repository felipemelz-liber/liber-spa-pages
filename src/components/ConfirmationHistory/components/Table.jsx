import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { CompleteTable } from 'liber-components';
import { DropDownContent, DropDownContentRow, EyeIcon, Global } from './Table.styles.jsx';
import { CONFIRMATION_HIST_COLUMNS, filtersPropTypes } from '../utils';
import { updatePagination } from '../actions/index.js';
import Filters from './filters/index.js';

export const ConfirmationHistoryTable = ({
  confirmations,
  handleFilter,
  pagination,
  onUpdatePagination,
  token,
  filters,
}) => {
  const { current, per, pages, count } = pagination;

  const listingStart = () => String((current - 1) * per + 1);
  const listingEnd = () => {
    const end = (current - 1) * per + per;
    if (count > end) return String(end);
    return String(count);
  };

  return (
    <>
      <Global />
      <CompleteTable
        columns={CONFIRMATION_HIST_COLUMNS}
        items={confirmations}
        filters={<Filters count={count} handleFilter={handleFilter} />}
        actions={({ details }) => (
          <DropDownContent>
            <DropDownContentRow href={details}>
              <EyeIcon />
              Visualizar Confirmação
            </DropDownContentRow>
          </DropDownContent>
        )}
        paginationProps={{
          currentPage: String(current),
          pageTotal: String(pages),
          onChangePage: target =>
            onUpdatePagination(token, { per, page: parseInt(target.value, 10), ...filters }),
          onClickNext: () => onUpdatePagination(token, { per, page: current + 1, ...filters }),
          onClickPrev: () => onUpdatePagination(token, { per, page: current - 1, ...filters }),
          hasListing: true,
          listingProps: {
            onSelectListing: selectedListing =>
              onUpdatePagination(token, {
                per: parseInt(selectedListing, 10),
                page: 1,
                ...filters,
              }),
            listing: String(per),
            total: String(count),
            listingLabelSingular: 'confirmação',
            listingLabelPlural: 'confirmações',
            showInterval: {
              start: listingStart(),
              end: listingEnd(),
            },
          },
          onKeyDownPage: () => null,
        }}
      />
    </>
  );
};

ConfirmationHistoryTable.propTypes = {
  confirmations: PropTypes.arrayOf(PropTypes.object),
  handleFilter: PropTypes.func,
  pagination: PropTypes.shape({
    current: PropTypes.number,
    per: PropTypes.number,
    pages: PropTypes.number,
    count: PropTypes.number,
  }),
  onUpdatePagination: PropTypes.func,
  token: PropTypes.string.isRequired,
  filters: filtersPropTypes.isRequired,
};

ConfirmationHistoryTable.defaultProps = {
  confirmations: [],
  handleFilter: () => null,
  pagination: {},
  onUpdatePagination: () => null,
};

const mapStateToProps = ({ confirmationHistory }) => ({
  confirmations: confirmationHistory.confirmations,
  pagination: confirmationHistory.pagination,
  token: confirmationHistory.token,
  filters: confirmationHistory.filters,
});

const mapDispatchToProps = {
  onUpdatePagination: updatePagination,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ConfirmationHistoryTable);
