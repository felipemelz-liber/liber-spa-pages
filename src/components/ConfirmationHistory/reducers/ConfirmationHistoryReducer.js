import Immutable from 'seamless-immutable';
import {
  CONFIRMATION_HISTORY_SET_CREDENTIALS,
  FETCH_CONFIRMATION_HISTORY_SUCCESS,
  FETCH_CONFIRMATION_HISTORY_REJECTED,
  FETCH_CONFIRMATION_HISTORY,
  CONFIRMATION_HISTORY_UPDATE_PAGINATION,
  CONFIRMATION_HISTORY_UPDATE_FILTERS,
} from '../actions/index.js';

export const initialState = Immutable({
  token: '',
  sacadoId: null,
  loadingAllPage: true,
  confirmations: [],
  error: '',
  pagination: {
    current: 1,
    previous: null,
    next: null,
    per: 10,
    pages: 1,
    count: 1,
  },
  filters: {
    'q[created_at_eq]': null,
    'q[created_at_gteq]': null,
    'q[created_at_lteq]': null,
  },
});

export default (state = initialState, action = {}) => {
  switch (action.type) {
    case CONFIRMATION_HISTORY_SET_CREDENTIALS:
      return state.merge({
        token: action.payload.token,
        sacadoId: action.payload.sacadoId,
      });
    case FETCH_CONFIRMATION_HISTORY_SUCCESS:
      return state.merge({
        confirmations: action.payload.confirmations,
        pagination: action.payload.pagination,
        loadingAllPage: false,
      });
    case FETCH_CONFIRMATION_HISTORY_REJECTED:
      return state.merge({ loadingAllPage: false, error: action.payload.message });
    case FETCH_CONFIRMATION_HISTORY:
      return state.merge({ loadingAllPage: true, error: '' });
    case CONFIRMATION_HISTORY_UPDATE_PAGINATION:
      return state.merge({
        pagination: {
          ...state.pagination,
          ...action.payload.pagination,
        },
      });
    case CONFIRMATION_HISTORY_UPDATE_FILTERS:
      return state.merge({
        filters: {
          ...action.payload.filters,
        },
      });
    default:
      return state;
  }
};
