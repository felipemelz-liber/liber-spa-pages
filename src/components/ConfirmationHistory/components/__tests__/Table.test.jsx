import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';
import { ConfirmationHistoryTable } from '../Table';
import { confirmationsMock } from '../../reducers/__tests__/ConfirmationHistoryReducer.test';

describe('Testing ConfirmationHistoryTable component', () => {
  configure({ adapter: new Adapter() });

  const props = {
    token: 'token',
    filters: {
      'q[created_at_eq]': '',
      'q[created_at_gteq]': '',
      'q[created_at_lteq]': '',
    },
    pagination: {
      current: 1,
      per: 10,
      pages: 1,
      count: 2,
    },
  };

  it('should render empty table correctly', () => {
    const wrapper = shallow(
      <ConfirmationHistoryTable confirmations={[]} handleFilter={() => null} {...props} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render table correctly', () => {
    const wrapper = shallow(
      <ConfirmationHistoryTable
        confirmations={confirmationsMock}
        handleFilter={() => null}
        {...props}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
