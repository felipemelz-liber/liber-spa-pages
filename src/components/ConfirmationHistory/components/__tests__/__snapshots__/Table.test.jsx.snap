// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ConfirmationHistoryTable component should render empty table correctly 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <CompleteTable
    actions={[Function]}
    allChecked={false}
    borderBottom={false}
    columns={
      Array [
        Object {
          "align": "center",
          "label": "Código",
          "name": "slug",
          "width": 12,
        },
        Object {
          "align": "right",
          "getCellValue": [Function],
          "label": "Valor Total",
          "name": "value",
          "width": 18,
        },
        Object {
          "align": "right",
          "label": "Data de Criação",
          "name": "created_at",
          "width": 18,
        },
        Object {
          "align": "left",
          "getCellValue": [Function],
          "label": "Contrato",
          "name": "contract",
          "width": 25,
        },
      ]
    }
    emptyState={<Empty />}
    filterBorderTop={true}
    filters={
      <Filters
        count={2}
        handleFilter={[Function]}
      />
    }
    headerBorderTop={true}
    hiddenColumns={Array []}
    isItemSelectable={[Function]}
    isItemSelected={[Function]}
    items={Array []}
    onSelect={[Function]}
    onSelectAll={null}
    paginationProps={
      Object {
        "currentPage": "1",
        "hasListing": true,
        "listingProps": Object {
          "listing": "10",
          "listingLabelPlural": "confirmações",
          "listingLabelSingular": "confirmação",
          "onSelectListing": [Function],
          "showInterval": Object {
            "end": "2",
            "start": "1",
          },
          "total": "2",
        },
        "onChangePage": [Function],
        "onClickNext": [Function],
        "onClickPrev": [Function],
        "onKeyDownPage": [Function],
        "pageTotal": "1",
      }
    }
    selectionControl={null}
    sortingProps={
      Object {
        "onSort": [Function],
        "sortedColumns": false,
        "sorting": Array [],
      }
    }
    tooltipContent={[Function]}
    weightedColumns={false}
  />
</Fragment>
`;

exports[`Testing ConfirmationHistoryTable component should render table correctly 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <CompleteTable
    actions={[Function]}
    allChecked={false}
    borderBottom={false}
    columns={
      Array [
        Object {
          "align": "center",
          "label": "Código",
          "name": "slug",
          "width": 12,
        },
        Object {
          "align": "right",
          "getCellValue": [Function],
          "label": "Valor Total",
          "name": "value",
          "width": 18,
        },
        Object {
          "align": "right",
          "label": "Data de Criação",
          "name": "created_at",
          "width": 18,
        },
        Object {
          "align": "left",
          "getCellValue": [Function],
          "label": "Contrato",
          "name": "contract",
          "width": 25,
        },
      ]
    }
    emptyState={<Empty />}
    filterBorderTop={true}
    filters={
      <Filters
        count={2}
        handleFilter={[Function]}
      />
    }
    headerBorderTop={true}
    hiddenColumns={Array []}
    isItemSelectable={[Function]}
    isItemSelected={[Function]}
    items={
      Array [
        Object {
          "contract": "http://demo.liber.local:3002/sacado/confirmacao_titulos",
          "created_at": "15/10/2020",
          "details": "#lala",
          "slug": "dcb234z",
          "value": 101056,
        },
        Object {
          "contract": "http://demo.liber.local:3002/sacado/confirmacao_titulos",
          "created_at": "12/08/2020",
          "details": "#lele",
          "slug": "dcb234f",
          "value": 531074,
        },
      ]
    }
    onSelect={[Function]}
    onSelectAll={null}
    paginationProps={
      Object {
        "currentPage": "1",
        "hasListing": true,
        "listingProps": Object {
          "listing": "10",
          "listingLabelPlural": "confirmações",
          "listingLabelSingular": "confirmação",
          "onSelectListing": [Function],
          "showInterval": Object {
            "end": "2",
            "start": "1",
          },
          "total": "2",
        },
        "onChangePage": [Function],
        "onClickNext": [Function],
        "onClickPrev": [Function],
        "onKeyDownPage": [Function],
        "pageTotal": "1",
      }
    }
    selectionControl={null}
    sortingProps={
      Object {
        "onSort": [Function],
        "sortedColumns": false,
        "sorting": Array [],
      }
    }
    tooltipContent={[Function]}
    weightedColumns={false}
  />
</Fragment>
`;
